Add component parameter for custom composition

ncovercash opened the pull request:
This PR adds support for a custom prop, `Component`, to be provided to override the default `<div>` used for each progress piece.  This will default to `<div>` when nothing is provided (current behavior), however, with this custom component users can more powerfully control the rendered progress pieces.

This allows a range of extra features, such as injecting special text, adding extra CSS, (in my case) wrapping each piece with a tooltip component, or any other customization.

Also, the PR includes a change to `package-lock.json` to reflect the package's current version.

ncovercash committed:
Add component parameter for custom composition

ncovercash commented:
An excerpt from react's typings:

```ts
type ElementType<P = any> =
    {
        [K in keyof JSX.IntrinsicElements]: P extends JSX.IntrinsicElements[K] ? K : never
    }[keyof JSX.IntrinsicElements] |
    ComponentType<P>;
type ComponentType<P = {}> = ComponentClass<P> | FunctionComponent<P>;
```

This means that users can provide any valid `ElementType`, either a builtin element (like `"div"`, `"p"`, or anything else), or any function/class component that accepts a `style`, `className`, and `children` prop.

N3XT191 commented:
Cool idea! I've tested a couple basic cases and there it worked, but I can't say for sure it will work for 100% of all possible cases, but I guess that's good enough 😅

What I don't like though is the prop name `Component`, I always use camelCase, and maybe something like `elementComponent` would be a bit more descriptive? If you change that, I'll merge this and publish a new version :)

And while we're at it, feel free to add the new prop to the README docs, those have to be updated manually I'm afraid, you can also include it in the second example. Otherwise I'll have to come up with my own example.

ncovercash commented:
Oops, for some reason I thought component props were supposed to be pascal case, but it seems I was completely wrong there. I'll update that and add to the readme soon!

ncovercash committed:
Allow custom props to be passed through to custom element components

ncovercash committed:
Fix import in readme example

ncovercash commented:
Ok, so I added a little more weirdness, in addition to renaming `Component` to `component` and adding to the README.  I added a generic that by default is `Record<string, never>` (empty object) that allows users to add custom properties to the `elements` array that will be forwarded to the component.

To take advantage of this, in plain JS users don't have to do anything; in TS they can add their own type as a generic onto the element, as shown in the README: `<MultiProgress<ExtraData> ... />`.  Then, all props defined in `ExtraData` will be passed down to their custom component for access/reuse, like in the example I put in the README (adds a custom `isBold` prop).  I also exported the prop types, just to make it easier for users to write their own components which accept them

N3XT191 commented:
I'm getting a weird type error when using your example in my demo app, it's most likely a Typescript and/or versioning problem I think. Can you check which versions you use? I have tried both the very latest react/typescript versions as well as the exact same versions used in your fork.

```Type '{ fontWeight: "bolder" | undefined; alignContent?: Property.AlignContent | undefined; alignItems?: Property.AlignItems | undefined; alignSelf?: Property.AlignSelf | undefined; ... 772 more ...; vectorEffect?: Property.VectorEffect | undefined; }' is not assignable to type 'Properties<string | number, string & {}>'.
  Types of property 'MozForceBrokenImageIcon' are incompatible.
    Type 'import("/Users/marc/Documents/PRIVAT/projects/temp/react-multi-progress/node_modules/csstype/index").Property.MozForceBrokenImageIcon | undefined' is not assignable to type 'import("/Users/marc/Documents/PRIVAT/projects/demo-multi-progress/node_modules/csstype/index").Property.MozForceBrokenImageIcon | undefined'.
      Type 'number & {}' is not assignable to type 'MozForceBrokenImageIcon | undefined'.
        Type 'number & {}' is not assignable to type '1'.
    19 | 		<div
    20 | 			className={props.className}
  > 21 | 			style={{
       | 			^^^^^
    22 | 				...props.style,
    23 | 				fontWeight: props.element.isBold ? "bolder" : undefined,
    24 | 			}}
```

N3XT191 commented:
I can get around the error by adding `as React.CSSProperties` as follows:

```
style={
				{
					...props.style,
					fontWeight: props.element.isBold ? "bolder" : undefined,
				} as React.CSSProperties
			}
```

but it's not the most elegant solution.

And even with this fix, it doesn't render the ProgressElements, I am trying to track down the exact problem but it looks like it doesn't create the divs properly with width/height etc. so they don't get rendered

EDIT: I think it's somewhere in the `createElementArray` function where props/styles/children or smth don't get passed on properly

ncovercash commented:
@N3XT191 oh weird, didn't realize what `glamor` was doing with the CSS.  It looks like glamor is returning a prop to be spread onto the component (something starting with data...) -- is there any reason this was chosen?  Just curious, since I hadn't seen something like this before (I'll update my PR to work with that soon, though)

ncovercash committed:
Fix style passing method

ncovercash commented:
ok, so changed the props to just be a Record<string, something>, that way the `data-something` attribute can be passed to the receiving component.  also added notes about spreading that on

ncovercash committed:
Fix any typing

N3XT191 commented:
Perfect, works now! (With the exception of `CustomComponent` in your example where you destructure `props.element` and `props.children` so you then have to use only `children` and `element`, at least in my configuration)

The reason why glamor was chosen is just because that's what I was using in projects back when I created this 2 years ago, if you feel the desire to replace it with smith more modern feel free, I don't really touch this code often enough to care too much 🙃

N3XT191 commented:
(Published as 1.3.0)

ncovercash commented:
No worries, I was just curious about it!  Glamor seems neat, just hadn't seen something like that before

ncovercash commented:
Thanks for the merge!
